kernel/nativevalues.ml

```diff
@@ -82,29 +82,33 @@ type symbols = symbol array
 
 let empty_symbols = [| |]
 
-
-let accumulate_tag = 0
-
 (** Unique pointer used to drive the accumulator function *)
 let ret_accu = Obj.repr (ref ())
 
 type accu_val = { mutable acc_atm : atom; acc_arg : Obj.t list }
 
+external register_accu : t -> unit = "coq_native_register_accu" [@@noalloc]
+external is_accu : t -> bool = "coq_native_is_accu" [@@noalloc]
+external is_closure : 'a -> bool = "coq_native_is_clos" [@@noalloc]
+
+(** TODO: do this more efficiently in C? *)
+(* Dummy wrapper to ensure the closure is not inlined by the compiler *)
```

Comment:
I do not understand why this would prevent any inlining. It seems better to use the old code and put the attribute `[@inline never]` on it:
```ocaml
let rec [@inline never] accumulate data x =
  if x == ret_accu then Obj.repr data
  else Obj.repr (accumulate { data with acc_arg = x :: data.acc_arg })
```
Another way to prevent inlining (if you do not trust the attribute) is to simply store the accumulator in a public mutable reference.

Reply:
This doesn't work because we will get partially applied functions, which have not the accumulator code. I need to have a function of arity one that is exactly the one performing accumulation for my trick to work.

Reply:
Indeed, my mistake. Still, I do not see how your code guarantees the uniqueness of the inner function. Perhaps something like the following would help, but I am not even sure OCaml guarantees that the `[@inline]` attribute is respected for inner functions.

```ocaml
let rec accumulate data =
  let [@inline never] accfun x =
    if x == ret_accu then Obj.repr data
    else
      let data = { data with acc_arg = x :: data.acc_arg } in
      Obj.repr (accumulate data).accfun in
  { accfun }
```

Here is another possibility, which is documented (or at least folklore enough to not be broken that easily):
```ocaml
let [@inline never] rec accumulate data = (); fun x ->
  if x == ret_accu then Obj.repr data
  else Obj.repr (accumulate { data with acc_arg = x :: data.acc_arg })
```

Reply:
I believe flambda explicitly breaks the second kind of trick. I witnessed that when observing the code generated by the proof monad with flambda.

Reply:
At least, with 4.12.0+flambda, it is compiled as expected, that is, `();fun` forces an explicit closure creation.

Reply:
Quoting Vincent Laviron:
> Flambda doesn’t try to coalesce a function returning a function into a function taking more arguments. Some transformations earlier in the compiler do that, but at that point `();` will not have been simplified yet so you should be fine.

Reply:
Other suggestion by Vincent:
```ocaml
let [@inline never] rec accumulate data = Sys.opaque_identity (fun x ->
  if x == ret_accu then Obj.repr data
  else Obj.repr (accumulate { data with acc_arg = x :: data.acc_arg }))
```